### PR TITLE
feat(tests): Add client and server test coverage to Gulp

### PR DESCRIPTION
**Pull request by hyperreality**
feat(tests): Add client and server test coverage to Gulp

This runs server-side test coverage using grunt-istanbul and sets karma
to run coverage tests, when gulp test:coverage is run.

I need a project maintainer to help integrate coveralls to this. After that is
done, Gulp should be 100% ready to replace Grunt.

Completes one of the #1175 milestones, also #1231 can be closed

### CODE REVIEW

**Comment by lirantal** [comment]
@hyperreality awesome, what do you need my help with?

**Comment by simison** [comment]
@hyperreality Could you change this, too: https://github.com/meanjs/mean/blob/master/.travis.yml#L41

**Comment by lirantal** [comment]
True, that will kick off the coveralls so we can see some results, and also maybe we should update the README/npm scripts in `package.json` for a coverage command.

**Comment by hyperreality** [comment]
@lirantal I wasn't quite sure how to integrate coveralls at the time, but I found a package `gulp-coveralls` so I'll have a go at implementing that with my next commit.

@simison yes I'll edit that with my next commit here, good catch.

File: package.json [hunk 1]
```diff
@@ -50,6 +50,7 @@
     "glob": "~7.0.0",
     "grunt": "~1.0.1",
     "grunt-cli": "~1.2.0",
+    "gulp-istanbul": "^1.1.1",
```
Comment (ilanbiala) [inline package.json]: `

Comment (hyperreality) [inline package.json]: @ilanbiala thanks, noted.

**Comment by hyperreality** [comment]
Okay this is pretty weird stuff. Here are the results of the grunt coverage task: https://coveralls.io/builds/7677048
And here are the results of the new gulp coverage task: https://coveralls.io/builds/7677185

Karma is testing a different set of client side files each time! The original grunt coverage test misses out client-side directives, controllers and services. On the other hand, the gulp coverage test misses out all 'config' directory files which were included in the grunt task.

I'll take another look at this tomorrow.

**Comment by lirantal** [comment]
Ok, keep us posted @hyperreality, this one is really important to merge when ready.

**Comment by hyperreality** [comment]
Okay, I think I cracked it. There was a bug with the previous Grunt coverage tasks never submitting client-side tests to coveralls. This new Gulp coverage task fixes that by merging the client and server side test reports and submitting them together. Thus the number of files being recorded for coverage has increased.

Should be ready now I believe.

**Comment by hyperreality** [comment]
I can't seem to get the checks kicking off on this one.

**Comment by mleanos** [comment]
@hyperreality Looks like there's a conflict with your version of the branch & *master*. Try re-basing your local *master* branch from upstream, and then re-basing your PR branch from your local *master*.

**Comment by hyperreality** [comment]
@mleanos Cheers, I didn't see those new commits coming in.

**Comment by lirantal** [comment]
@hyperreality how is it now?
seems that the coverage is now back to testing only the unit tests, not e2e coverage as well. Can you take a look?

**Comment by hyperreality** [comment]
@lirantal I'm not sure what's going on. My personal builds are getting coverage for all tests: https://coveralls.io/builds/7697944

**Comment by lirantal** [comment]
Very odd.
I'll anyway merge this in and will appreciate if you can track it further to figure out. Maybe the travis build process is the issue.

**Pull request merged**